airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSchemas.java

```diff
@@ -59,4 +92,137 @@ public static void mutateTypeToArrayStandard(final JsonNode jsonNode) {
+    return new HashSet<>(traverseJsonSchemaWithCollector(obj, (node, path) -> {
+      if (predicate.test(node)) {
+        return Optional.of(path);
+      } else {
+        return Optional.empty();
+      }
+    }));
+  }
+
+  /**
+   * Recursive, depth-first implementation of { @link JsonSchemas#traverseJsonSchema(final JsonNode
+   * jsonNode, final BiConsumer<JsonNode, List<String>> consumer) }. Takes path as argument so that
+   * the path can be passsed to the consumer.
+   *
+   * @param jsonSchemaNode - jsonschema object to traverse.
+   * @param path - path from the first call of traverseJsonSchema to the current node.
+   * @param consumer - consumer to be called at each node. it accepts the current node and the path to the node from the
+   *        root of the object passed at the root level invocation
+   */
+  // todo (cgardens) - replace with easier to understand traversal logic from SecretsHelper.
+  private static void traverseJsonSchemaInternal(final JsonNode jsonSchemaNode,
+                                         final String path,
+                                         final BiConsumer<JsonNode, String> consumer) {
+    if (!jsonSchemaNode.isObject()) {
+      throw new IllegalArgumentException(String.format("json schema nodes should always be object nodes. path: %s actual: %s", path, jsonSchemaNode));
+    }
+
+    consumer.accept(jsonSchemaNode, path);
+    // if type is missing assume object. not official JsonSchema, but it seems to be a common
+    // compromise.
+    final List<String> nodeTypes = getTypeOrObject(jsonSchemaNode);
+
+    for (final String nodeType : nodeTypes) {
+      switch (nodeType) {
+        case ARRAY_TYPE -> {
+          // todo (cgardens) - remove magic string.
+          final String newPath = path + "[*]";
+          // hit every node.
+          traverseJsonSchemaInternal(jsonSchemaNode.get(JSON_SCHEMA_ITEMS_KEY), newPath, consumer);
+//          jsonNode.forEach(node -> traverseJsonSchema(jsonSchemaNode.get(JSON_SCHEMA_ITEMS_KEY), node, newPath, consumer));
+        }
+        case OBJECT_TYPE -> {
+          final Optional<String> comboKeyWordOptional = getKeywordIfComposite(jsonSchemaNode);
+          if (jsonSchemaNode.has(JSON_SCHEMA_PROPERTIES_KEY)) {
+            for (final Iterator<Entry<String, JsonNode>> it = jsonSchemaNode.get(JSON_SCHEMA_PROPERTIES_KEY).fields(); it.hasNext();) {
+              final Entry<String, JsonNode> child = it.next();
+              // todo (cgardens) - remove magic string.
+              final String newPath = path + "." + child.getKey();
+              traverseJsonSchemaInternal(child.getValue(), newPath, consumer);
+            }
+          } else if (comboKeyWordOptional.isPresent()) {
+            for (final JsonNode arrayItem : jsonSchemaNode.get(comboKeyWordOptional.get())) {
+              traverseJsonSchemaInternal(arrayItem, path, consumer);
+            }
+          } else {
```

Comment:
How does it works with an airbyte secret, do we need to handle it. I think that it will have a key: airbyte_secret: true.

Reply:
we can add a validation that airbyte_secret is never an object. we should treat it as invalid.

Reply:
(i will replace this confusing logic so it is clearer)